src/java/org/apache/cassandra/service/accord/AccordJournalTable.java

```diff
@@ -0,0 +1,183 @@
+    private final ColumnFamilyStore cfs;
+
+    private final ColumnMetadata recordColumn;
+    private final ColumnMetadata versionColumn;
+
+    private final Journal<K, V> journal;
+    private final KeySupport<K> keySupport;
+    private final int keySerializationVersion;
+
+
+    public AccordJournalTable(Journal<K, V> journal, KeySupport<K> keySupport, int keySerializationVersion)
+    {
+        this.journal = journal;
+        this.cfs = Keyspace.open(AccordKeyspace.metadata().name).getColumnFamilyStore(AccordKeyspace.JOURNAL);
+        this.recordColumn = cfs.metadata().getColumn(ColumnIdentifier.getInterned("record", false));
+        this.versionColumn = cfs.metadata().getColumn(ColumnIdentifier.getInterned("user_version", false));
+        this.keySupport = keySupport;
+        this.keySerializationVersion = keySerializationVersion;
+    }
+
+    /**
+     * Perform a read from Journal table, followed by the reas from all journal segments.
+     * <p>
+     * When reading from journal segments, skip descriptors that were read from the table.
+     */
+    public void readAll(K id, RecordConsumer<K> consumer)
+    {
+        journal.readAll(id,
+                        new RecordConsumer<>()
+                        {
+                            LongHashSet visitedDescriptors = new LongHashSet();
+                            boolean collectDescriptors = true;
+
+                            @Override
+                            public void init()
+                            {
+                                readAllFromTable(id, this);
+                                collectDescriptors = false;
+                            }
+
+                            @Override
+                            public void accept(long segment, int position, K key, ByteBuffer buffer, IntHashSet hosts, int userVersion)
+                            {
+                                // To avoid duplicates, we need to filter out segment descriptors that were visited while reading from the table
+                                if (collectDescriptors)
+                                    visitedDescriptors.add(segment);
+                                else if (visitedDescriptors.contains(segment))
+                                    return;
+
+                                consumer.accept(segment, position, key, buffer, hosts, userVersion);
+                            }
+                        });
+    }
+
+    protected void readAllFromTable(K key, RecordConsumer<K> onEntry)
+    {
+        List<Ref<SSTableReader>> refs = null;
+        try
+        {
+            EntrySerializer.EntryHolder<K> into = new EntrySerializer.EntryHolder<>();
+            DecoratedKey pk = cfs.decorateKey(JournalKey.Support.serialize(keySupport, key, keySerializationVersion));
+            Collection<SSTableReader> sstables = cfs.select(View.select(SSTableSet.LIVE, pk)).sstables;
```

Comment:
I think we should be using `CFS.selectAndReference` here? Is there a reason why we cannot? Otherwise I think this code is possibly incorrect? Think skipping sstables that have just been compacted away.